src/scheduler/work.rs

```diff
@@ -0,0 +1,45 @@
+use super::worker::*;
+use super::*;
+use crate::mmtk::MMTK;
+use crate::vm::VMBinding;
+use std::any::{type_name, TypeId};
+
+pub trait Work<C: Context>: 'static + Send + Sync {
+    fn do_work(&mut self, worker: &mut Worker<C>, context: &'static C);
+    #[inline]
+    fn do_work_with_stat(&mut self, worker: &mut Worker<C>, context: &'static C) {
+        let stat = worker
+            .stat
+            .measure_work(TypeId::of::<Self>(), type_name::<Self>());
+        self.do_work(worker, context);
+        stat.end_of_work(&mut worker.stat);
+    }
+}
+
+impl<VM: VMBinding> PartialEq for Box<dyn Work<VM>> {
+    /// Address comparison. Two different work instances cannot have the same address
+    #[allow(clippy::vtable_address_comparisons)]
+    fn eq(&self, other: &Self) -> bool {
```

Comment:
Thanks for pointing this out. I did some tests of address comparison of boxed structs / trait-objects. It turns out that my assumption here is incorrect. If a struct has no fields, then there will be no heap allocation. The boxed value will hold a fake pointer (e.g. 0x01) instead of a pointer into the heap.

The purpose of this address comparison is for fast work-packet comparison. Otherwise comparing (or calculating hash value for) work-packets with large data buffers will be extremely slow. I'll add a unique work-id for each instance of work-packet and use that for fast comparison.

Reply:
Fixed. Now each work is assigned with a unique ID when it is added to a bucket. And the priority-queue in buckets will use this ID for fast work-packet comparison.